src/System.Net.Http/src/System/Net/Http/HttpRequestMessage.cs

```diff
@@ -27,6 +27,9 @@ public class HttpRequestMessage : IDisposable
         private bool _disposed;
         private IDictionary<String, Object> _properties;
 
+        // In case of request redirects, set to true to use credential cache.
+        internal bool _useCredentialCache;
```

Comment:
Is this really the right place to store this?  Basically, we're adding an internal field (public within the entire System.Net.Http.dll assembly) just to fix an implementation issue with the managed handler.

Doing this breaks the SOLID principles of object oriented design.

Reply:
In WinHttpHandler, this sort of thing is tracked not in the HttpRequestMessage but rather in the per-message state object that is used during the lifetime of the request processing in WinHttpHandler.

We should use a similar technique for storing transient state like this.

Reply:
See: https://github.com/dotnet/corefx/blob/master/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestState.cs

Reply:
The handlers execute multiple requests with the same settings, and this particular credential setting will be specific to a particular request that is getting redirected. Since the setting is specific to the particular request object, this seemed clean. Do you have other suggestions on how to fix this?

Reply:
If you look at the design of the WinHttpRequestState object and how it is used, you will see that is created for each request that goes thru the handler.  So, it is specific to a "particular request".  You can see that the state variable snapshots the critical handler properties including credentials.  Then, when necessary, those variables are mutated (i.e. for redirection the creds are cleared).  This is a cleaner design than adding an a transient state variable to the HttpRequestMessage.

cc: @stephentoub

Reply:
Hmm, that requires a rewrite of the managed handler to have a similar design to winhttphandler. In case of winhttphandler, there's just a single handler, and a request state object that stores all the request specific settings. And then for each request, there's a new request state object, and this request state object is used to set the intermittent settings on the winhttphandler, which translates to setting the options on the native handle.

In case of managedhandler, it has several handlers that are initialized once, and each request is executed with SendAsync on these handlers. Even with having a new request state object similar to the winhttp design, there's still the issue of passing down this request state object to the managed handlers. The ```SendAsync(HttpRequestMessage,CancellationToken)``` is derived from ```HttpMessageHandler``` and this cannot be modified to take the new request state object.

This reduces to having to find a way to reset a handle's state (authentication handler in this case) for every request. Is there a way to do this with the current managed handler design?